simulator/func_sim/func_sim.cpp

```diff
@@ -2,15 +2,29 @@
  * func_sim.cpp - extremely simple simulator
  * Copyright 2018 MIPT-MIPS
  */
- 
+
 #include "func_sim.h"
 #include <kernel/kernel.h>
+#include <iostream>
+#include <sstream>
+#include <stdexcept>
+
 
 template <typename ISA>
-FuncSim<ISA>::FuncSim( Endian endian, bool log)
+typename FuncSim<ISA>::HandleTrapStrMap FuncSim<ISA>::handle_trap_str_map = {
+    { "stop",         HandleTrapMode::STOP         },
+    { "stop_on_halt", HandleTrapMode::STOP_ON_HALT },
+    { "ignore",       HandleTrapMode::IGNORE       },
+};
+
+template <typename ISA>
+FuncSim<ISA>::FuncSim( Endian endian, bool log, const std::string &trap_mode, bool trap_critical, bool trap_verbose)
     : Simulator( log)
     , imem( endian)
     , kernel( Kernel::create_dummy_kernel())
+    , handle_trap_mode(handle_trap_str_map.at(trap_mode))
+    , handle_trap_critical(trap_critical)
+    , handle_trap_verbose(trap_verbose)
```

Comment:
Could you please explain why we have 3 control variables instead of a single variable, at least for ctor arguments?

Reply:
There are 4 incompatible behaviors described by the string, two options set by switch. Should I make them 6 in one enum?

Reply:
I don't care much about the internals, but it is more convenient to have 1 string argument instead of 1 string argument + 2 Boolean arguments. Boolean arguments are evil since you don't understand their meanings on the caller side (at least in C++)